src/main/kotlin/org/rust/ide/highlight/RsHighlightExitPointsHandlerFactory.kt

```diff
@@ -0,0 +1,105 @@
+            return RsHighlightExitPointsHandler(editor, file, target)
+        }
+        if (elementType == IDENTIFIER) {
+            val parent = target.parent
+            if (parent is RsMacroCall && (parent.referenceName == "panic" || parent.referenceName == "try")) {
+                return RsHighlightExitPointsHandler(editor, file, target)
+            }
+        }
+        if (testForExprBeforeBlock(target)) {
+            return RsHighlightExitPointsHandler(editor, file, target)
+        }
+        return null
+    }
+
+    private fun testForExprBeforeBlock(target: PsiElement): Boolean {
+        var context: PsiElement? = target
+        while (context != null) {
+            val parent = context.parent
+            if (context is RsExpr && parent is RsBlock) {
+                return true
+            }
+            context = parent
+        }
+        return false
+    }
+
+    private class RsHighlightExitPointsHandler(editor: Editor, file: PsiFile, var target: PsiElement) : HighlightUsagesHandlerBase<PsiElement>(editor, file) {
+        override fun getTargets() = listOf(target)
+
+        override fun selectTargets(targets: MutableList<PsiElement>?, selectionConsumer: Consumer<MutableList<PsiElement>>?) {
+            selectionConsumer?.consume(targets)
+        }
+
+        private fun getFunctionOrLambda(target: PsiElement): PsiElement? {
+            var context: PsiElement? = target
+            while (context != null) {
+                if (context is RsFunction || context is RsLambdaExpr) {
+                    return context
+                }
+                context = context.parent
+            }
+            return null
+        }
+
+        override fun computeUsages(targets: MutableList<PsiElement>?) {
+            val function: PsiElement? = getFunctionOrLambda(target)
+            function?.acceptChildren(object : RsVisitor() {
+                override fun visitElement(element: PsiElement?) {
+                    element?.acceptChildren(this)
+                }
+
+                override fun visitFunction(o: RsFunction) {}
+                override fun visitLambdaExpr(o: RsLambdaExpr) {}
+                override fun visitRetExpr(o: RsRetExpr) = addOccurrence(o)
+                override fun visitTryExpr(o: RsTryExpr) = addOccurrence(o)
+
+                override fun visitTryMacroArgument(o: RsTryMacroArgument) {
+                    val macroCall = o.parent as? RsMacroCall ?: return
+                    addOccurrence(macroCall)
+                }
+
+                override fun visitFormatMacroArgument(o: RsFormatMacroArgument) {
+                    val macroCall = o.parent as? RsMacroCall ?: return
+                    if (macroCall.referenceName == "panic") {
+                        addOccurrence(macroCall)
+                    }
+                }
+
+                override fun visitExpr(o: RsExpr) {
+                    if (o.parent is RsBlock) {
+                        addOccurrence(o)
```

Comment:
I had a look at how IntelliJ handles such cases for Groovy, which allows to omit `return`s the same way as Rust does.

It only highlights exit points when you put the cursor to a `return` word and doesn't highlight them when the cursor is inside an expression, even if this expression's result returns from the function:

<img width="131" alt="screen shot 2017-07-24 at 14 25 54" src="https://user-images.githubusercontent.com/2101250/28521396-237496b0-707c-11e7-97de-9d859a3b6a26.png">

<img width="133" alt="screen shot 2017-07-24 at 14 26 07" src="https://user-images.githubusercontent.com/2101250/28521406-2c7865c0-707c-11e7-91f6-40874ba7b360.png">

<img width="131" alt="screen shot 2017-07-24 at 14 26 23" src="https://user-images.githubusercontent.com/2101250/28521413-31e6e1bc-707c-11e7-96c1-d05983b28e53.png">

It looks as a reasonable approach to me.